tests/system/test_dbapi.py

```diff
@@ -544,6 +570,245 @@ def _insert_row(self, i):
+        connection works."""
+
+        method_count_interceptor = dbapi_database._method_count_interceptor
+        method_count_interceptor.reset()
+        self._insert_row(1)
+        self._insert_row(2)
+        self._cursor.execute("SELECT * FROM contacts")
+        self._cursor.fetchall()
+        dbapi_database._method_abort_interceptor.set_method_to_abort(COMMIT_METHOD)
+        self._conn.commit()
+        dbapi_database._method_abort_interceptor.reset()
+        assert method_count_interceptor._counts[COMMIT_METHOD] == 2
+        assert method_count_interceptor._counts[EXECUTE_STREAMING_SQL_METHOD] == 6
+
+        method_count_interceptor = dbapi_database._method_count_interceptor
+        method_count_interceptor.reset()
+        self._insert_row(3)
+        self._insert_row(4)
+        self._cursor.execute("SELECT * FROM contacts")
+        self._cursor.fetchall()
+        dbapi_database._method_abort_interceptor.set_method_to_abort(COMMIT_METHOD)
+        self._conn.commit()
+        dbapi_database._method_abort_interceptor.reset()
+        assert method_count_interceptor._counts[COMMIT_METHOD] == 2
+        assert method_count_interceptor._counts[EXECUTE_STREAMING_SQL_METHOD] == 6
+
+        self._cursor.execute("SELECT * FROM contacts")
+        got_rows = self._cursor.fetchall()
+        assert len(got_rows) == 4
+
+    @pytest.mark.noautofixt
+    def test_abort_retry_multiple_cursors(self, shared_instance, dbapi_database):
+        """Test that retry works when multiple cursors are involved in the transaction."""
+
+        try:
+            conn = Connection(shared_instance, dbapi_database)
+            cur = conn.cursor()
+            cur.execute(
+                """
+                CREATE TABLE Singers (
+                    SingerId     INT64 NOT NULL,
+                    Name    STRING(1024),
+                ) PRIMARY KEY (SingerId)
+            """
+            )
+            cur.execute(
+                """
+            INSERT INTO contacts (contact_id, first_name, last_name, email)
+            VALUES (1, 'first-name', 'last-name', 'test.email@domen.ru')
+            """
+            )
+            cur.execute(
+                """
+            INSERT INTO Singers (SingerId, Name)
+            VALUES (1, 'first-name')
+            """
+            )
+            cur.execute(
+                """
+            INSERT INTO contacts (contact_id, first_name, last_name, email)
+            VALUES (2, 'first-name', 'last-name', 'test.email@domen.ru')
+            """
+            )
+            cur.execute(
+                """
+            INSERT INTO Singers (SingerId, Name)
+            VALUES (2, 'first-name')
+            """
+            )
+            conn.commit()
+
+            cur1 = conn.cursor()
+            cur1.execute("SELECT * FROM contacts")
+            cur2 = conn.cursor()
+            cur2.execute("SELECT * FROM Singers")
+            row1 = cur1.fetchone()
```

Comment:
Note (not for this pull request, but something that we should look into in a future pull request): `cursor.fetchone`, `cursor.fetchmany` and `cursor.fetchall` can all also return an `Aborted` error, even when we are halfway through the results. We should verify that that situation is also handled gracefully. That means that either of the following should happen:
1. The method just propagates the `Aborted` error to the application and the transaction fails.
2. OR: The cursor continues from where it was.

So assume the following query

```
select *
from singers
```

which returns:

```
(1, 'FirstName1', 'LastName1')
(2, 'FirstName2', 'LastName2')
(3, 'FirstName3', 'LastName3')
```

The application then calls:

```
// Returns row 1
cursor.fetchone()

// Raises an Aborted error.
This should either be propagated to the application, OR internally retried and then return row 2.
cursor.fetchone()

// Returns row 3 (or is never reached if the previous `fetchone` just propagated the Aborted error.
cursor.fetchone() 
```

The reason that `fetchone` (and other fetch methods) can return `Aborted` is that `ExecuteStreamingSql` returns a stream of `PartialResultSet`. This stream can break __at any moment__ with an `Aborted` error if the transaction has been aborted by Cloud Spanner.

Reply:
The code takes care of retry when Aborted exception happens during any of the fetch call but I have not been able to add tests for it

Reply:
Will add the tests later. Create a bug  for this https://buganizer.corp.google.com/issues/319404848

Reply:
ACK